### PR TITLE
Clean shutdown of nodeipam controller

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref https://github.com/kubernetes/kubernetes/issues/108483

```release-note
NONE
```

/kind cleanup
/priority important-longterm
/sig network

/assign @aojea

### CODE REVIEW

File: pkg/controller/nodeipam/node_ipam_controller.go [hunk 1]
```diff
@@ -151,7 +159,9 @@ func (nc *Controller) Run(stopCh <-chan struct{}) {
 		return
 	}
 
-	if nc.allocatorType != ipam.IPAMFromClusterAllocatorType && nc.allocatorType != ipam.IPAMFromCloudAllocatorType {
+	if nc.allocatorType == ipam.IPAMFromClusterAllocatorType || nc.allocatorType == ipam.IPAMFromCloudAllocatorType {
+		go nc.legacyIPAM.Run(stopCh)
```
Comment (aojea) [inline pkg/controller/nodeipam/node_ipam_controller.go:163]: this was tricky

File: test/integration/framework/etcd.go [hunk 2]
```diff
@@ -196,9 +196,9 @@ func EtcdMain(tests func() int) {
 	checkNumberOfGoroutines := func() (bool, error) {
 		// Leave some room for goroutines we can not get rid of
 		// like k8s.io/klog/v2.(*loggingT).flushDaemon()
-		// TODO(#108483): Reduce this number once we address the
-		//   couple remaining issues.
-		if dg := runtime.NumGoroutine() - before; dg <= 9 {
+		// TODO(#108483): Figure out if we can reduce this
+		//   further (ideally down to zero).
+		if dg := runtime.NumGoroutine() - before; dg <= 4 {
```
Comment (aojea) [inline test/integration/framework/etcd.go:201]: this is incredible

**Comment by aojea** [comment]
/lgtm

Thanks

**Comment by wojtek-t** [comment]
Just fixed gofmt - retagging.

**Pull request merged**